AzureStorageBlob/Source/DataStructures/StorageStreamDownloader.swift

```diff
@@ -0,0 +1,435 @@
+
+        if let requestId = requestId { headerParams["x-ms-client-request-id"] = requestId }
+        if let leaseId = leaseAccessConditions?.leaseId { headerParams["x-ms-lease-id"] = leaseId }
+        if let encryptionKey = cpk?.value { headerParams["x-ms-encryption-key"] = String(data: encryptionKey, encoding: .utf8) }
+        if let encryptionKeySHA256 = cpk?.hash { headerParams["x-ms-encryption-key-sha256"] = encryptionKeySHA256 }
+        if let encryptionAlgorithm = cpk?.algorithm { headerParams["x-ms-encryption-algorithm"] = encryptionAlgorithm }
+        if let ifModifiedSince = modifiedAccessConditions?.ifModifiedSince { headerParams[.ifModifiedSince] = ifModifiedSince.rfc1123Format }
+        if let ifUnmodifiedSince = modifiedAccessConditions?.ifUnmodifiedSince { headerParams[.ifUnmodifiedSince] = ifUnmodifiedSince.rfc1123Format }
+        if let ifMatch = modifiedAccessConditions?.ifMatch { headerParams[.ifMatch] = ifMatch }
+        if let ifNoneMatch = modifiedAccessConditions?.ifNoneMatch { headerParams[.ifNoneMatch] = ifNoneMatch }
+
+        // Construct and send request
+        let request = client.request(method: HttpMethod.GET,
+                                   url: url,
+                                   queryParams: queryParams,
+                                   headerParams: headerParams)
+        let context: [String: AnyObject] = [
+            ContextKey.allowedStatusCodes.rawValue: [200, 206] as AnyObject,
+        ]
+        client.run(request: request, context: context) { result, httpResponse in
+            switch result {
+            case let .failure(error):
+                completion(.failure(error), httpResponse)
+            case let .success(data):
+                guard let data = data else {
+                    completion(.failure(AzureError.general("Blob unexpectedly contained no data.")), httpResponse)
+                    return
+                }
+                let headers = httpResponse.headers
+                if let contentMD5 = headers["Content-MD5"] {
+                    let dataHash = data.md5
+                    guard contentMD5 == dataHash else {
+                        let error = AzureError.general("Block MD5 \(dataHash) did not match \(contentMD5).")
+                        completion(.failure(error), httpResponse)
+                        return
+                    }
+                }
+                if let contentCRC64 = headers["Content-CRC64"] {
+                    let dataHash = data.crc64
+                    guard contentCRC64 == dataHash else {
+                        let error = AzureError.general("Block CRC64 \(dataHash) did not match \(contentCRC64).")
+                        completion(.failure(error), httpResponse)
+                        return
+                    }
+                }
+                let decryptedData = self.decrypt(data)
+                do {
+                    let handle = try self.openFileForWriting()
```

Comment:
Should we move the call to `closeFile()` up into a `defer` block here?

Reply:
Specifically, since opening the file itself is an operation that might fail, it would be something like
`defer { handle?.closeFile() }`

Reply:
Yes, good catch! In Swift the best time to put a defer statement is right after the one that makes it necessary in the first place. Fixed.